artemis-commons/src/main/java/org/apache/activemq/artemis/utils/SizeAwareMetric.java

```diff
@@ -0,0 +1,236 @@
+
+   public final long addSize(final int delta) {
+      return addSize(delta, false);
+   }
+
+   public final long addSize(final int delta, final boolean sizeOnly) {
+
+      if (delta == 0) {
+         if (logger.isDebugEnabled()) {
+            logger.debug("SizeAwareMetric ignored with size 0", new Exception("trace"));
+         }
+         return sizeUpdater.get(this);
+      }
+
+      changeFlag(NOT_USED, FREE);
+
+      if (onSizeCallback != null) {
+         try {
+            onSizeCallback.add(delta, sizeOnly);
+         } catch (Throwable e) {
+            logger.warn(e.getMessage(), e);
+         }
+      }
+
+      long currentSize = sizeUpdater.addAndGet(this, delta);
+
+      long currentElements;
+      if (sizeOnly) {
+         currentElements = elementsUpdater.get(this);
+      } else if (delta > 0) {
+         currentElements = elementsUpdater.incrementAndGet(this);
+      } else {
+         currentElements = elementsUpdater.decrementAndGet(this);
+      }
+
+      if (delta > 0) {
+         checkOver(currentElements, currentSize);
+      } else { // (delta < 0)
+         checkUnder(currentElements, currentSize);
+      }
+
+      return currentSize;
+   }
+
+   public void setMax(long maxSize, long lowerMarkSize, long maxElements, long lowerMarkElements) {
+      this.maxSize = maxSize;
+      this.lowerMarkSize = lowerMarkSize;
+      this.maxElements = maxElements;
+      this.lowerMarkElements = lowerMarkElements;
+      long currentSize = sizeUpdater.get(this);
+      long currentElements = elementsUpdater.get(this);
+      checkOver(currentElements, currentSize);
+      checkUnder(currentElements, currentSize);
+   }
+
+   private void checkUnder(long currentElements, long currentSize) {
+      if (sizeEnabled) {
+         if (currentSize <= lowerMarkSize && changeFlag(OVER_SIZE, FREE)) {
+            // checking if we need to switch from OVER_SIZE to OVER_ELEMENTS, to avoid calling under needless
+            if (!(elementsEnabled && currentElements >= maxElements && changeFlag(FREE, OVER_ELEMENTS))) {
+               under();
+            }
+            return; // we must return now as we already checked for the elements portion
+         }
+      }
+
+      if (elementsEnabled) {
+         if (currentElements <= lowerMarkElements && changeFlag(OVER_ELEMENTS, FREE)) {
+            // checking if we need to switch from OVER_ELEMENTS to OVER_SIZE, to avoid calling under needless
+            if (!(sizeEnabled && currentSize >= maxSize && changeFlag(FREE, OVER_SIZE))) {
+               // this is checking the other side from size (elements).
+               // on this case we are just switching sides and we should not fire under();
+               under();
+            }
+            return; // this return here is moot I know. I am keeping it here for consistence with the size portion
+                    //  and in case eventually further checks are added on this method, this needs to be reviewed.
+         }
+      }
```

Comment:
Is it ok for it to fire _over_ repeatedly without ever firing _under_  inbetween, as this would then seem possible now?

Reply:
it's the opposite... it will only fire under if this wasn't achieved...

think of this as 

```java
            if (sizeEnabled && currentSize >= maxSize && changeFlag(FREE, OVER_SIZE)) {
               // nothing to be done as it switch over.
             } else {
               // this is checking the other side from size (elements).
               // on this case we are just switching sides and we should not fire under();
               under();
            }
```


I actually had it like this first and I then added the ! to the check

Reply:
it would still be fine .. although I don't think that's the case.

Reply:
This stops it firing 'under' here if the _other_ threshold is still over when the first goes down again, yes. However, my point was around that it still sets the state to FREE first, so could something else happen that makes it realise the other threshhold is breached, which will set that as the state, and possibly fire _over_ again having never fired _under_. Though in retrospect I guess it will fire _under_ here though if it fails to _changeFlag(FREE, <other>)_. So it will fire, just maybe not when expected (could be after it fired _over_ in the other check).